### ISSUE TITLE
Difference between the loops in the train and train_step methods

### ISSUE THREAD

**Issue by ajesujoba** [pytorch/fairseq#1666]
In the `train(args, trainer, task, epoch_itr): ` method in the `train.py` file there is a loop which is `for samples in progress:` where the `train_step()` is invoked using `log_output = trainer.train_step(samples)`.  I am a little bit confused as there is another loop in the `train_step(self, samples, dummy_batch=False, raise_oom=False):` method `for i, sample in enumerate(samples):` found in the `Trainer `class. 

I assume the `samples` in the `for samples in progress:` are batches or is it the whole training set? If they are batches, why do we need the `for i, sample in enumerate(samples):`? Apologies if my assumption isn't correct. Thanks

**Comment by myleott** [comment]
It has to do with gradient accumulation. When `update_freq > 1` then `samples` will actually be a list of batches due to the `GroupedIterator`: https://github.com/pytorch/fairseq/blob/810e61ab3ffba169a47d7e72538e082d2467f2b4/fairseq_cli/train.py#L160

`Trainer.train_step` then iterates over these and accumulates the gradients.

**Comment by ajesujoba** [comment]
Thanks for you response! But when update_freq = 1,  `samples` is a list of a batch right? 

I guess instead of using `for samples in progress` It is possible to use `for samples in itr`?  I’m working on a self-supervised MT setting where batches aren’t available at ones and gradient computation is not done at once on the batches(as they are generated online or on the run).

**Comment by myleott** [comment]
> But when update_freq = 1, samples is a list of a batch right?

Yes when update_freq=1 then samples is a list with a single batch, something like:
```
samples = [
  {
    'id': torch.Tensor(...),
    'net_input': {
      (...)
    }
  }
]
```

> it is possible to use `for samples in itr`

Sure that works. But I don't think it will buy you anything, since progress is just a wrapper around iterator that adds a progress bar.

**Comment by ajesujoba** [comment]
Thank you once again for the clarification. I do appreciate. Do you think it is reasonable for me to have a `progress` object for each batch in my case(where batches are not always ready at the same time - and training is done on the available batches as soon as they are available)?

**Comment by myleott** [comment]
> So the idea would be to train on the batches as they are available

I’m not sure I understand. Are you generating the data in some other (external) process? And then reading data on the fly? You’re welcome to hack train.py however you need. Ultimately you just need to call train_step with a batch object, it doesn’t really matter where it comes from. You don’t need to use `itr` or `progress` at all.

**Comment by ajesujoba** [comment]
In my case, the sentence pairs are not available at the beginning, but we have some article pairs that we expect to contain sentence pairs(similar sentence). So our model is to extract the possible sentence pairs based on some similarity threshold of their embeddings, train on those sentence pairs as they are available during every epoch. So we are generating/reading the data by an external process during every epoch and training on the sentence pairs created during each epoch.

**Comment by ajesujoba** [comment]
Hi @myleott sorry I’m asking my question here but I think it is related to this issue I raised earlier. 

I’m training a NMT with a batch size of 50 (max length of source and target = 60) and update_freq=1. Please do you have an idea of what possible(most appropriate) value I should initialize my learning rate based on the batch size and update frequency?
